marlowe-runtime/web-server/Language/Marlowe/Runtime/Web/Server/REST/Contracts.hs

```diff
@@ -160,18 +199,67 @@ put
 put eb contractId body = withEvent eb Put \ev -> do
   addField ev $ PutId contractId
   contractId' <- fromDTOThrow (badRequest' "Invalid contract id value") contractId
+
   loadContract contractId' >>= \case
     Nothing -> throwError $ notFound' "Contract not found"
     Just (Left (TempTx _ Unsigned Tx.ContractCreated{txBody})) -> do
-      textEnvelope <- fromDTOThrow (badRequest' "Invalid body value") body
-      addField ev $ Body textEnvelope
-      tx <- either (const $ throwError $ badRequest' "Invalid body text envelope content") pure $ deserialiseFromTextEnvelope (AsTx AsBabbage) textEnvelope
-      unless (getTxBody tx == txBody) $ throwError (badRequest' "Provided transaction body differs from the original one")
+      let
+        -- `<|>` gives me error here
+        req :: Maybe (Either (Cardano.Tx BabbageEra) (ShelleyTxWitness BabbageEra))
+        req = case fromDTO (Left body) of
+          Just res -> pure res
+          Nothing -> fromDTO (Right body)
```

Comment:
We could probably choose the decoder based on the `teType` (I'm not 100% sure on the exact strings but I think these are right?)

```suggestion
        req :: Maybe (Either (Cardano.Tx BabbageEra) (ShelleyTxWitness BabbageEra))
        req = case (teType body) of
          "Tx BabbageEra" -> Left <$> fromDTO body
          "TxBody BabbageEra" -> Right <$> fromDTO body
          _ -> Nothing
```